spark/src/main/scala/org/apache/hadoop/fs/versioned/VersionedFileSystem.scala

```diff
@@ -0,0 +1,138 @@
+
+  override def create(
+      f: Path,
+      permission: FsPermission,
+      overwrite: Boolean,
+      bufferSize: Int,
+      replication: Short,
+      blockSize: Long,
+      progress: Progressable): FSDataOutputStream =
+    baseFS.create(baseVersionedPath(f), permission, overwrite, bufferSize, replication, blockSize, progress)
+
+  override def append(f: Path, bufferSize: Int, progress: Progressable): FSDataOutputStream =
+    baseFS.append(baseVersionedPath(f), bufferSize, progress)
+
+  override def rename(src: Path, dst: Path): Boolean =
+    baseFS.rename(baseVersionedPath(src), baseVersionedPath(dst))
+
+  override def delete(f: Path, recursive: Boolean): Boolean =
+    baseFS.delete(baseVersionedPath(f), recursive)
+
+  override def listStatus(f: Path): Array[FileStatus] =
+    baseFS.listStatus(setScheme(baseURI.getScheme)(f)) map { fileStatus =>
+      new FileStatus(
+        fileStatus.getLen,
+        fileStatus.isDirectory,
+        fileStatus.getReplication.toInt,
+        fileStatus.getBlockSize,
+        fileStatus.getModificationTime,
+        versionedPath(fileStatus.getPath)
+      )
+    } filter { fileStatus =>
+      fileStatus.getPath.toUri.getSchemeSpecificPart.contains(version) ||
+      fileStatus.isDirectory
+    }
+
+  override def setWorkingDirectory(new_dir: Path): Unit =
+    baseFS.setWorkingDirectory(baseVersionedPath(new_dir))
+
+  override def getWorkingDirectory: Path =
+    setScheme(baseURI.getScheme)(baseFS.getWorkingDirectory)
+
+  override def mkdirs(f: Path, permission: FsPermission): Boolean =
+    baseFS.mkdirs(baseVersionedPath(f), permission)
+
+  override def getFileStatus(f: Path): FileStatus = {
+    val fileStatus = baseFS.getFileStatus(setScheme(baseURI.getScheme)(f))
+
+    new FileStatus(
+      fileStatus.getLen,
+      fileStatus.isDirectory,
+      fileStatus.getReplication.toInt,
+      fileStatus.getBlockSize,
+      fileStatus.getModificationTime,
+      versionedPath(fileStatus.getPath)
+    )
+  }
+
+  private lazy val appendVersion: Path => Path = path => {
+    val uri = path.toUri
+
+    val sp = mappings.foldLeft(uri.getSchemeSpecificPart) {
+      case (acc, (partition, versionedPartition)) if !acc.contains(versionedPartition) && acc.contains(partition) =>
+        acc.replace(partition, versionedPartition)
+      case (acc, _) => acc
+    }
+
+    new Path(new URI(s"${uri.getScheme}:$sp"))
+  }
+
+  private lazy val setScheme: String => Path => Path =
+    scheme => path => new Path(new URI(s"$scheme:${path.toUri.getSchemeSpecificPart}"))
+
+  private lazy val toSchemeAndVersion: String => Path => Path = scheme => setScheme(scheme) andThen appendVersion
+
+  private lazy val baseVersionedPath: Path => Path = path => toSchemeAndVersion(baseURI.getScheme)(path)
+  private lazy val versionedPath: Path => Path = path => setScheme(VersionedFileSystem.SCHEME)(path)
+
+  override protected def checkPath(path: Path): Unit = super.checkPath(baseVersionedPath(path))
+}
```

Comment:
There are quite a few methods on `Filesystem` that aren't overridden here, that probable should be. If you have a look at this example proxy: https://github.com/apache/hive/blob/master/shims/common/src/main/java/org/apache/hadoop/fs/ProxyFileSystem.java, you'll see it overrides a number of other methods.

I think we want to ensure that the versioning layer is completely water tight in that it can not leak the underlying filesystem to the client in any scenario. So I think it would be worth going through the other methods and ensuring they're all handled (i.e. path translated both ways).

Reply:
I have added the extra methods that `ProxyFileSystem` does have. Although it does break the tests, it worked without them. For now, I am leaving them in as comments for ease of testing once against a read s3 filesystem.
I did try to run test against a local s3 back-end, direct to S3AFS and through this proxy, and get all sorts of errors around renaming files, permissions etc. In other words, adding those methods did not help. The pb is somewhere else (and is probably just test plumbing).